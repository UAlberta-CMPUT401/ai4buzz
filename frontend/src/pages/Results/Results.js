import React from 'react';
import FacialAnalysisResults from '../../components/FacialAnalysisResults/FacialAnalysisResults';
import ImageClassificationResults from '../../components/ImageClassificationResults/ImageClassificationResults';
import ObjectDetectionResults from '../../components/ObjectDetectionResults/ObjectDetectionResults';
import SentimentClassification from '../../components/SentimentClassification/SentimentClassification';
import styles from './Results.module.css';

const Results = ({ results, images }) => {
  const downloadFile = ({ featureAnalysisResults, fileName, fileType }) => {
    const blob = new Blob([featureAnalysisResults], { type: fileType });

    const a = document.createElement('a');
    a.download = fileName;
    a.href = window.URL.createObjectURL(blob);
    const clickEvt = new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
    });
    a.dispatchEvent(clickEvt);
    a.remove();
  };

  const exportToJson = (e) => {
    e.preventDefault();
    downloadFile({
      featureAnalysisResults: JSON.stringify(results),
      fileName: 'results.json',
      fileType: 'text/json',
    });
  };

  return (
    <div className={styles.resultsPage}>
      <div>
        <h1 className={styles.resultTitle}>Analysis Results</h1>

        <button type="button" onClick={exportToJson}>
          Export to JSON
        </button>

        <div className={styles.resultListContainer}>
          {results.feature_analysis_results.map((imageFeatures, idx) => {
            return (
              <div className={styles.imageFeatureResult} key={idx}>
                <h4>{imageFeatures.id}</h4>
                <div className={styles.imageContainer}>
                  <img
                    className={styles.image}
                    src={`data:image/jpeg;base64,${imageFeatures.object_detection.processes_bounding_boxes_image_as_base64_string}`}
                    alt="results"
                  />
                </div>
                <div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>Color Analysis</div>
                  </div>
                  <div>
                    {JSON.stringify(imageFeatures.color_scheme_analysis)}
                  </div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>
                      Object Detection
                    </div>
                  </div>
                  <div>
                    <ObjectDetectionResults
                      results={imageFeatures.object_detection}
                    />
                  </div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>
                      Image Classification
                    </div>
                  </div>
                  <div>
                    <ImageClassificationResults
                      results={imageFeatures.image_classification}
                    />
                  </div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>
                      Sentiment Classification
                    </div>
                  </div>
                  <div>
                    <SentimentClassification
                      results={imageFeatures.sentiment_analysis.degrees}
                    />
                  </div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>
                      Text Recognition
                    </div>
                  </div>
                  <div>
                    {imageFeatures.text_recognition === '' ? (
                      'No text detected'
                    ) : (
                      <div>
                        Extracted:
                        <div className="value">
                          {imageFeatures.text_recognition}
                        </div>
                      </div>
                    )}
                  </div>
                  <div className={styles.gradient}>
                    <div className={styles.analysisHeader}>Facial Analysis</div>
                  </div>
                  <div>
                    <FacialAnalysisResults
                      facialResults={imageFeatures?.face_analysis?.analysis}
                    />
                  </div>
                </div>
              </div>
            );
          })}
        </div>
      </div>
    </div>
  );
};

export default Results;
