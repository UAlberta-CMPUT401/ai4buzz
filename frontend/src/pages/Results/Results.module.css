.resultsPage {
  padding: 0 0 100px 0;
  display: flex;
  justify-content: center;
}

.imageContainer {
  display: flex;
  justify-content: center;
}

.imageFeatureResult {
  margin-bottom: 5rem;
}

.gradient {
  margin-top: 1.5rem;
  background: linear-gradient(
    27deg,
    rgba(255, 110, 208, 1) 0%,
    rgba(165, 43, 255, 1) 50%,
    rgba(0, 237, 255, 1) 100%
  );
  padding-bottom: 2px;
}

.analysisHeader {
  font-size: 1.3rem;
  font-weight: 500;
  background-color: #0e121c;
}

.resultTitle {
  margin-left: 1.5rem;
}

.resultListContainer {
  width: 512px;
}

.image {
  max-height: 512px;
  max-width: 512px;
}

@media (max-width: 712px) {
  .resultListContainer {
    width: 100%;
    padding: 0 1rem;
  }
}
