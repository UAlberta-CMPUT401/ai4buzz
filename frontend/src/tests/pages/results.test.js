import React from 'react';
import Results from '../../pages/Results/Results';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

const mockData = {
  feature_analysis_results: {
    image_1: {
      color_scheme_analysis: {
        count: 5,
        colors: [
          { red: 230, green: 226, blue: 216, proportion: 0.4253989209432653 },
          { red: 9, green: 24, blue: 38, proportion: 0.16556650478570417 },
          { red: 172, green: 165, blue: 158, proportion: 0.15257673629119714 },
          { red: 112, green: 96, blue: 78, proportion: 0.09586900439840229 },
          { red: 37, green: 33, blue: 23, proportion: 0.07738628479823256 },
        ],
      },
      object_detection: {
        person: { freq: 1, confidences: [0.9984568357467651] },
        laptop: {
          freq: 2,
          confidences: [0.9387264847755432, 0.6356411576271057],
        },
        cup: { freq: 1, confidences: [0.9060431718826294] },
        processes_bounding_boxes_image_as_base64_string: 'base 64 image string',
      },
      sentiment_analysis: {
        'sentiment_array[neg,neu,pos]': [
          0.5963839888572693, 0.13197453320026398, 0.2716415226459503,
        ],
        degrees: { Negative: '59.64%', Neutral: '13.20%', Postive: '27.16%' },
      },
      image_classification: {
        'cellular telephone': 0.6936329007148743,
        notebook: 0.19310395419597626,
        'hand-held computer': 0.04424683377146721,
        'web site': 0.021844087168574333,
        television: 0.014548316597938538,
      },
      text_recognition: '',
    },
  },
  collage_image_string: 'base 64 string',
};

const files = [
  new File(['image-bits...'], 'chucknorris.png', {
    type: 'image/png',
  }),
];

global.URL.createObjectURL = jest.fn();

it('Renders Results page correctly', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Results results={mockData} images={files} />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
