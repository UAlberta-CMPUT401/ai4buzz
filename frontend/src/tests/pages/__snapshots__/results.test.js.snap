// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Results page correctly 1`] = `
<div
  className="resultsPage"
>
  <div>
    <h1
      className="resultTitle"
    >
      Analysis Results
    </h1>
    <button
      onClick={[Function]}
      type="button"
    >
      Export to JSON
    </button>
    <div
      className="resultListContainer"
    >
      <div
        className="imageFeatureResult"
      >
        <h4 />
        <div
          className="imageContainer"
        >
          <img
            alt="results"
            className="image"
            src="data:image/jpeg;base64,base 64 image string"
          />
        </div>
        <div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Color Analysis
            </div>
          </div>
          <div>
            <div>
              <div>
                RGB Values and Proportion
              </div>
              <div
                style={
                  Object {
                    "backgroundColor": "rgb(230, 226, 216)",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "padding": "0.5rem",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    230
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    226
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    216
                  </span>
                </span>
                <span>
                  <span
                    style={
                      Object {
                        "float": "right",
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    42.540
                    %
                  </span>
                </span>
              </div>
              <div
                style={
                  Object {
                    "backgroundColor": "rgb(9, 24, 38)",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "padding": "0.5rem",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    9
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    24
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    38
                  </span>
                </span>
                <span>
                  <span
                    style={
                      Object {
                        "float": "right",
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    16.557
                    %
                  </span>
                </span>
              </div>
              <div
                style={
                  Object {
                    "backgroundColor": "rgb(172, 165, 158)",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "padding": "0.5rem",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    172
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    165
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    158
                  </span>
                </span>
                <span>
                  <span
                    style={
                      Object {
                        "float": "right",
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    15.258
                    %
                  </span>
                </span>
              </div>
              <div
                style={
                  Object {
                    "backgroundColor": "rgb(112, 96, 78)",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "padding": "0.5rem",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    112
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    96
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    78
                  </span>
                </span>
                <span>
                  <span
                    style={
                      Object {
                        "float": "right",
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    9.587
                    %
                  </span>
                </span>
              </div>
              <div
                style={
                  Object {
                    "backgroundColor": "rgb(37, 33, 23)",
                  }
                }
              >
                <span
                  style={
                    Object {
                      "padding": "0.5rem",
                    }
                  }
                >
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    37
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    33
                    , 
                  </span>
                  <span
                    style={
                      Object {
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    23
                  </span>
                </span>
                <span>
                  <span
                    style={
                      Object {
                        "float": "right",
                        "mixBlendMode": "difference",
                      }
                    }
                  >
                    7.739
                    %
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Object Detection
            </div>
          </div>
          <div>
            <div
              className="objectDetectionResultsData"
            >
              <div>
                person
              </div>
              <div>
                <div
                  className="value"
                >
                  99.846
                  %
                </div>
              </div>
            </div>
            <div
              className="objectDetectionResultsData"
            >
              <div>
                laptop
              </div>
              <div>
                <div
                  className="value"
                >
                  93.873
                  %
                </div>
                <div
                  className="value"
                >
                  63.564
                  %
                </div>
              </div>
            </div>
            <div
              className="objectDetectionResultsData"
            >
              <div>
                cup
              </div>
              <div>
                <div
                  className="value"
                >
                  90.604
                  %
                </div>
              </div>
            </div>
          </div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Image Classification
            </div>
          </div>
          <div>
            <div>
              <div
                className="imageClassificationResultsData"
              >
                <div>
                  cellular telephone
                </div>
                <div
                  className="value"
                >
                  69.363
                  %
                </div>
              </div>
              <div
                className="imageClassificationResultsData"
              >
                <div>
                  notebook
                </div>
                <div
                  className="value"
                >
                  19.310
                  %
                </div>
              </div>
              <div
                className="imageClassificationResultsData"
              >
                <div>
                  hand-held computer
                </div>
                <div
                  className="value"
                >
                  4.425
                  %
                </div>
              </div>
              <div
                className="imageClassificationResultsData"
              >
                <div>
                  web site
                </div>
                <div
                  className="value"
                >
                  2.184
                  %
                </div>
              </div>
              <div
                className="imageClassificationResultsData"
              >
                <div>
                  television
                </div>
                <div
                  className="value"
                >
                  1.455
                  %
                </div>
              </div>
            </div>
          </div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Sentiment Classification
            </div>
          </div>
          <div>
            <div>
              <div
                className="sentimentClassificationData"
              >
                <div>
                  Negative
                </div>
                <div
                  className="value"
                >
                  59.64%
                </div>
              </div>
              <div
                className="sentimentClassificationData"
              >
                <div>
                  Neutral
                </div>
                <div
                  className="value"
                >
                  13.20%
                </div>
              </div>
              <div
                className="sentimentClassificationData"
              >
                <div>
                  Postive
                </div>
                <div
                  className="value"
                >
                  27.16%
                </div>
              </div>
            </div>
          </div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Text Recognition
            </div>
          </div>
          <div>
            No text detected
          </div>
          <div
            className="gradient"
          >
            <div
              className="analysisHeader"
            >
              Facial Analysis
            </div>
          </div>
          <div>
            <div>
              <div>
                <div
                  className="faceResults"
                >
                  <div>
                    Age
                  </div>
                  <div
                    className="value"
                  >
                    22
                  </div>
                </div>
                <div
                  className="faceResults"
                >
                  <div>
                    Gender
                  </div>
                  <div
                    className="value"
                  >
                    Man
                  </div>
                </div>
                <div
                  className="faceResults"
                >
                  <div>
                    Dominant Emotion
                  </div>
                  <div
                    className="value"
                  >
                    happy
                  </div>
                </div>
                <div>
                  Emotion
                </div>
                <div
                  className="tabbedData"
                >
                  <div
                    className="faceResults"
                  >
                    <div>
                      angry
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      disgust
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      fear
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      happy
                    </div>
                    <div
                      className="value"
                    >
                      99.986
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      sad
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      surprise
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      neutral
                    </div>
                    <div
                      className="value"
                    >
                      0.014
                      %
                    </div>
                  </div>
                </div>
                <div
                  className="faceResults"
                >
                  <div>
                    Dominant Race
                  </div>
                  <div
                    className="value"
                  >
                    white
                  </div>
                </div>
                <div>
                  Race
                </div>
                <div
                  className="tabbedData"
                >
                  <div
                    className="faceResults"
                  >
                    <div>
                      Asian
                    </div>
                    <div
                      className="value"
                    >
                      0.001
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      Indian
                    </div>
                    <div
                      className="value"
                    >
                      0.018
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      Middle Eastern
                    </div>
                    <div
                      className="value"
                    >
                      11.545
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      White
                    </div>
                    <div
                      className="value"
                    >
                      86.629
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      Latino Hispanic
                    </div>
                    <div
                      className="value"
                    >
                      1.807
                      %
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      Black
                    </div>
                    <div
                      className="value"
                    >
                      0.000
                      %
                    </div>
                  </div>
                </div>
                <div>
                  Region
                </div>
                <div
                  className="tabbedData"
                >
                  <div
                    className="faceResults"
                  >
                    <div>
                      x
                    </div>
                    <div
                      className="value"
                    >
                      17
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      y
                    </div>
                    <div
                      className="value"
                    >
                      16
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      w
                    </div>
                    <div
                      className="value"
                    >
                      169
                    </div>
                  </div>
                  <div
                    className="faceResults"
                  >
                    <div>
                      h
                    </div>
                    <div
                      className="value"
                    >
                      169
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
